## 2. Conhecimento Inicial dos Dados

### 2.1 Dimensões do dataset

In [ ]:
n_linhas, n_colunas = df.shape
print(f"Quantidade de linhas (avaliações): {n_linhas}")
print(f"Quantidade de colunas (variáveis): {n_colunas}")


Quantidade de linhas (avaliações): 24113
Quantidade de colunas (variáveis): 7


O dataset possui o número de linhas exibido acima, condizente com a
descrição do projeto de aproximadamente 24 mil avaliações. Cada linha é uma avaliação individual e
cada uma das colunas representa um atributo dessa avaliação ou do produto avaliado.

### 2.2 Primeiras observações

In [ ]:
df.head(10)


,product_name,overall_rating,no_ratings,no_reviews,rating,title,review
0,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,5,Perfect product!,"Loved it, it's my first MacBook that I earned from my hardwork 🥺❤️"
1,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,5,Fabulous!,Battery lasted longer than my first relationship (2 days).\nListening to Ari...
2,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,5,Fabulous!,Such a great deal.. very happy with the performance and battery life..Origio...
3,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,4,Delightful,"Awesome build quality and very good display, battery and camera. Still new t..."
4,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,5,Awesome,"When i ordered and came to know about seller review,i was scared that in wha..."
5,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,5,Super!,Super product
6,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,5,Super!,Go for it..its awesome
7,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,5,Mind-blowing purchase,"Best , best and best 🫶🏻👑🍎"
8,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,5,Just wow!,Its really very good and compact device.
9,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,5,Brilliant,"Superb built quality, Amazing performance and display quality...\n\nGot it a..."


In [ ]:
df.sample(5, random_state=RANDOM_STATE)


,product_name,overall_rating,no_ratings,no_reviews,rating,title,review
7228,ASUS Vivobook 15 Intel Core i5 12th Gen 1235U - (8 GB/512 GB S...,4.2,"5,789",310,4,Wonderful,Good product
1510,Apple MacBook AIR Apple M2 - (8 GB/512 GB SSD/Mac OS Monterey)...,4.7,"15,210",900,5,Perfect product!,"Loved it, it's my first MacBook that I earned from my hardwork 🥺❤️"
19793,HP 14s Intel Core i3 11th Gen 1115G4 - (8 GB/512 GB SSD/Window...,4.3,"12,747",954,5,Highly recommended,Overall nice product for home and students
20996,HP Chromebook Intel Celeron Dual Core N4500 - (4 GB/64 GB EMMC...,3.7,"1,384",133,5,Super!,Very very good product nice
2004,"ASUS Vivobook 15, with Backlit Keyboard, Intel Core i3 12th Ge...",4.3,"9,156",698,5,Simply awesome,Very good products


Observando as primeiras linhas e uma amostra aleatória, nota-se que várias
avaliações pertencem ao mesmo produto (mesmo `product_name`, `overall_rating`, `no_ratings` e
`no_reviews` repetidos), o que é esperado, já que essas três últimas colunas são atributos **do
produto**, não da avaliação individual. Os campos `rating`, `title` e `review` variam a cada linha,
pois são específicos de cada cliente.

### 2.3 Informações das colunas

In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24113 entries, 0 to 24112
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   product_name    24113 non-null  object 
 1   overall_rating  24113 non-null  float64
 2   no_ratings      24113 non-null  object 
 3   no_reviews      24113 non-null  object 
 4   rating          24113 non-null  int64  
 5   title           24113 non-null  object 
 6   review          24113 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 1.3+ MB


In [ ]:
info_df = pd.DataFrame({
    "coluna": df.columns,
    "tipo": df.dtypes.astype(str).values,
    "não_nulos": df.notna().sum().values,
    "nulos": df.isna().sum().values,
})
info_df


,coluna,tipo,não_nulos,nulos
0,product_name,object,24113,0
1,overall_rating,float64,24113,0
2,no_ratings,object,24113,0
3,no_reviews,object,24113,0
4,rating,int64,24113,0
5,title,object,24113,0
6,review,object,24113,0


Todas as colunas foram lidas pelo pandas, como será detalhado na seção de
Qualidade dos Dados, as colunas `no_ratings` e `no_reviews` foram importadas como texto (`object`)
porque seus valores contêm separador de milhar (vírgula), o que impede a leitura direta como número.
Isso será corrigido na seção 3.5.

### 2.4 Classificação das variáveis

Com base no significado de cada coluna (e não apenas no tipo bruto lido pelo pandas), classificamos
as variáveis da seguinte forma:

- **Numéricas:**
  - `rating` (inteira, 1 a 5) — variável-alvo do projeto, de natureza **ordinal** (existe uma ordem
    entre as classes), tratada como multiclasse na etapa de classificação;
  - `overall_rating` (contínua) — nota média do produto;
  - `no_ratings` (contínua/discreta, após conversão) — quantidade de notas do produto;
  - `no_reviews` (contínua/discreta, após conversão) — quantidade de avaliações escritas do produto.
- **Categóricas:**
  - `product_name` — variável nominal com muitas categorias (um valor por modelo de notebook);
  - `title` — título curto da avaliação; tem características híbridas (categórica/textual), pois é
    texto livre mas concentrado em um vocabulário relativamente pequeno de expressões (ex.: "Wonderful",
    "Terrific", "Just okay").
- **Textuais:**
  - `review` — texto livre, principal variável preditora do projeto futuro de NLP;
  - `title` — também analisado sob a ótica textual (tamanho, vocabulário), além de categórica.
- **Temporais:**
  - **Não existe nenhuma variável de data/hora no dataset.** Não há coluna como `data_da_avaliacao`
    ou `timestamp`. Portanto, **não é possível realizar nenhuma análise de série temporal** (evolução
    das notas ao longo do tempo, sazonalidade etc.) nesta etapa. Essa é uma limitação do dataset, que
    será registrada na conclusão.

O código abaixo apenas formaliza essa classificação em estruturas de dados que serão reutilizadas ao
longo do notebook.

In [ ]:
colunas_numericas_brutas = ["rating", "overall_rating"]
colunas_numericas_a_converter = ["no_ratings", "no_reviews"]
colunas_categoricas = ["product_name", "title"]
colunas_textuais = ["review", "title"]
colunas_temporais = []  # não existem no dataset

print("Numéricas (já corretas):", colunas_numericas_brutas)
print("Numéricas (a converter):", colunas_numericas_a_converter)
print("Categóricas:", colunas_categoricas)
print("Textuais:", colunas_textuais)
print("Temporais:", colunas_temporais if colunas_temporais else "Nenhuma variável temporal está presente no dataset.")


Numéricas (já corretas): ['rating', 'overall_rating']
Numéricas (a converter): ['no_ratings', 'no_reviews']
Categóricas: ['product_name', 'title']
Textuais: ['review', 'title']
Temporais: Nenhuma variável temporal está presente no dataset.


### 2.5 Estatísticas descritivas

Antes de calcular estatísticas para `no_ratings` e `no_reviews`, é necessário convertê-las para
numérico (elas chegam como texto por causa da vírgula separadora de milhar, ex.: `"15,210"`). Essa
conversão será formalmente documentada na seção 3.5 (Tipos Inadequados); aqui criamos versões
auxiliares (`_num`) apenas para permitir o cálculo das estatísticas desta seção, sem alterar ainda o
DataFrame original.

In [ ]:
no_ratings_num = df["no_ratings"].astype(str).str.replace(",", "", regex=False).astype(float)
no_reviews_num = df["no_reviews"].astype(str).str.replace(",", "", regex=False).astype(float)

print("Exemplo de conversão 'no_ratings':")
print(df["no_ratings"].head(3).tolist(), "->", no_ratings_num.head(3).tolist())


Exemplo de conversão 'no_ratings':
['15,210', '15,210', '15,210'] -> [15210.0, 15210.0, 15210.0]


#### 2.5.1 Variáveis numéricas

In [ ]:
numericas_df = pd.DataFrame({
    "rating": df["rating"],
    "overall_rating": df["overall_rating"],
    "no_ratings": no_ratings_num,
    "no_reviews": no_reviews_num,
})

desc_num = numericas_df.describe().T
desc_num["amplitude"] = desc_num["max"] - desc_num["min"]
desc_num["mediana"] = numericas_df.median()
desc_num = desc_num[["count", "mean", "mediana", "std", "min", "25%", "50%", "75%", "max", "amplitude"]]
desc_num.columns = ["contagem", "média", "mediana", "desvio_padrão", "mínimo", "Q1", "Q2 (mediana)", "Q3", "máximo", "amplitude"]
desc_num.round(2)


,contagem,média,mediana,desvio_padrão,mínimo,Q1,Q2 (mediana),Q3,máximo,amplitude
rating,24113.0,4.21,5.0,1.18,1.0,4.0,5.0,5.0,5.0,4.0
overall_rating,24113.0,4.19,4.2,0.23,3.3,4.1,4.2,4.3,5.0,1.7
no_ratings,24113.0,2712.26,1569.0,3155.61,1.0,660.0,1569.0,3475.0,15729.0,15728.0
no_reviews,24113.0,224.71,133.0,251.35,1.0,62.0,133.0,310.0,2171.0,2170.0


A coluna `rating` (nota individual da avaliação) varia entre 1 e 5, com
média acima de 4, sugerindo predominância de notas altas (confirmado com detalhes na Análise
Univariada). `overall_rating` tem uma amplitude bem menor, pois é uma média agregada por produto
(portanto suaviza valores extremos). `no_ratings` e `no_reviews` apresentam desvio-padrão muito maior
que a média, uma primeira indicação de forte assimetria à direita (poucos produtos muito populares
concentrando a maior parte das avaliações), isso será investigado com mais detalhe na seção de
outliers e na análise univariada.

#### 2.5.2 Variáveis categóricas

In [ ]:
for col in colunas_categoricas:
    n_cat = df[col].nunique()
    print(f"Coluna '{col}': {n_cat} categorias distintas")

print()
print("Top 10 categorias de 'product_name' (frequência e percentual):")
freq_produto = df["product_name"].value_counts().head(10)
pct_produto = (df["product_name"].value_counts(normalize=True) * 100).head(10)
display(pd.DataFrame({"frequência": freq_produto, "percentual (%)": pct_produto.round(2)}))

print()
print("Top 10 categorias de 'title' (frequência e percentual):")
freq_title = df["title"].value_counts().head(10)
pct_title = (df["title"].value_counts(normalize=True) * 100).head(10)
display(pd.DataFrame({"frequência": freq_title, "percentual (%)": pct_title.round(2)}))


Coluna 'product_name': 365 categorias distintas
Coluna 'title': 213 categorias distintas

Top 10 categorias de 'product_name' (frequência e percentual):


,frequência,percentual (%)
product_name,,
ASUS Vivobook 15 Intel Core i5 12th Gen 1235U - (8 GB/512 GB S...,400,1.66
CHUWI Intel Celeron Dual Core 11th Gen N4020 - (8 GB/256 GB SS...,400,1.66
Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/Mac OS Monterey)...,400,1.66
HP Backlit Intel Core i5 12th Gen 1235U - (8 GB/512 GB SSD/Win...,330,1.37
ASUS TUF Gaming F15 - AI Powered Gaming Intel Core i5 11th Gen...,314,1.30
ASUS Vivobook Go 15 OLED AMD Ryzen 3 Quad Core 7320U - (8 GB/5...,310,1.29
DELL Intel Core i3 12th Gen 1215U - (8 GB/512 GB SSD/Windows 1...,302,1.25
ASUS Vivobook Go 15 OLED AMD Ryzen 5 Quad Core 7520U - (8 GB/5...,300,1.24
ZEBRONICS Pro Series Z Intel Core i5 12th Gen 1235U - (8 GB/51...,300,1.24



Top 10 categorias de 'title' (frequência e percentual):


,frequência,percentual (%)
title,,
Wonderful,1266,5.25
Super!,829,3.44
Best in the market!,813,3.37
Fabulous!,812,3.37
Highly recommended,803,3.33
Classy product,802,3.33
Terrific purchase,799,3.31
Terrific,783,3.25
Perfect product!,769,3.19


`product_name` possui um número elevado de categorias (um valor distinto
por modelo de notebook), com forte concentração de avaliações em poucos produtos muito vendidos/
avaliados (ex.: o produto mais frequente já responde por uma fatia expressiva do total, calculada
acima). `title` tem número de categorias bem menor, e é dominado por expressões curtas e positivas
("Wonderful", "Super!", "Best in the market!"), o que já sinaliza - antes mesmo de olhar para
`rating` - a tendência de avaliações majoritariamente positivas na plataforma.

#### 2.5.3 Variáveis textuais

In [ ]:
df["review_chars"] = df["review"].astype(str).str.len()
df["review_words"] = df["review"].astype(str).str.split().apply(len)
df["title_chars"] = df["title"].astype(str).str.len()
df["title_words"] = df["title"].astype(str).str.split().apply(len)

txt_desc = pd.DataFrame({
    "review_chars": df["review_chars"].describe(),
    "review_words": df["review_words"].describe(),
    "title_chars": df["title_chars"].describe(),
    "title_words": df["title_words"].describe(),
}).T
txt_desc["mediana"] = pd.Series({
    "review_chars": df["review_chars"].median(),
    "review_words": df["review_words"].median(),
    "title_chars": df["title_chars"].median(),
    "title_words": df["title_words"].median(),
})
txt_desc = txt_desc[["count", "mean", "mediana", "std", "min", "25%", "50%", "75%", "max"]]
txt_desc.round(2)


,count,mean,mediana,std,min,25%,50%,75%,max
review_chars,24113.0,96.91,46.0,122.46,2.0,17.0,46.0,123.0,510.0
review_words,24113.0,17.47,8.0,22.18,1.0,3.0,8.0,22.0,108.0
title_chars,24113.0,12.55,11.0,5.38,4.0,9.0,11.0,16.0,60.0
title_words,24113.0,1.90,2.0,0.95,1.0,1.0,2.0,2.0,13.0


O texto da `review` tem, em média, o número de caracteres e palavras
mostrado acima, mas com desvio-padrão maior que a média - sinal de que a distribuição do tamanho das
avaliações é bastante assimétrica, com muitas avaliações curtas e algumas bem longas (o máximo chega
a centenas de caracteres). O `title`, por natureza, é bem mais curto (poucas palavras), consistente
com seu papel de "resumo" da avaliação.